# Data Science Pandas

## Tasks Today:

0) <b>Pre-Work</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Numpy Random Sampling

1) <b>Pandas</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Importing <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Tabular Data Structures <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - from_dict() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - read_csv() <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) <b>In-Class Exercise #1</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Accessing Data <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Indexing <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - df.loc <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - keys() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Slicing a DataFrame <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Built-In Methods <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - head() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - tail() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - shape <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - describe() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - sort_values() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - .columns <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) <b>In-Class Exercise #2</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; g) Filtration <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Conditionals <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Subsetting <br>
 &nbsp;&nbsp;&nbsp;&nbsp; h) Column Transformations <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Generating a New Column w/Data <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - User Defined Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; i) Aggregations <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - groupby() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - Type of groupby() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - mean() <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - groupby() w/Multiple Columns <br>
 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - drop_duplicates() <br>

## Numpy Random Sampling

In [159]:
import numpy as np

# A single call generates a single random number
print('Here is a random number: %s' % np.random.uniform())

# You can also pass some bounds
print('Here is a random number between 0 and 1 Million: %s' % np.random.uniform(0, 1e6))

# You can also generate a bunch of random numbers all at once
print('Here are 3 random numbers between 0 and 1 Million: %s' % np.random.uniform(0, 1e6, 3))

# Even matrices and tensors with weird shapes
print('Here is a 3x3 matrix of random numbers between 0 and 1 Million: \n %s' % np.random.uniform(0, 1e6, (3, 3)))

# instead of float values, generating random integers
print('Using Random Integer: %s' % np.random.randint(0, 10, 4))

Here is a random number: 0.03585250796283923
Here is a random number between 0 and 1 Million: 657402.6603573285
Here are 3 random numbers between 0 and 1 Million: [ 56674.63255861 715184.63945082 490989.05824656]
Here is a 3x3 matrix of random numbers between 0 and 1 Million: 
 [[854746.66890949 794354.13442878 903442.47438462]
 [187079.33301497 727109.69689612 197829.60761833]
 [105582.68457849 111773.6589589  802084.86309839]]
Using Random Integer: [8 1 5 4]


## Pandas <br>

<p>Pandas is a flexible data analysis library built on top of NumPy that is excellent for working with tabular data. It is currently the de-facto standard for Python-based data analysis, and fluency in Pandas will do wonders for your productivity and frankly your resume. It is one of the fastest ways of getting from zero to answer in existence. </p>

<ul>
    <li>Pandas is a Python module, written in C. The Pandas module is a high performance, highly efficient, and high level data analysis library. It allows us to work with large sets of data called dataframes.</li>
    <li>Series is a one-dimensional labeled array capable of holding data of any type (integer, string, float, python objects, etc.)</li>
    <li>Dataframe = Spreadsheet (has column headers, index, etc.)</li>
</ul>

### Importing

In [160]:
import pandas as pd

# always use pd, standard for data science

### Tabular data structures <br>
<p>The central object of study in Pandas is the DataFrame, which is a tabular data structure with rows and columns like an excel spreadsheet. The first point of discussion is the creation of dataframes both from native Python dictionaries, and text files through the Pandas I/O system.</p>

In [161]:
names = ['Alice',
         'Bob',
         'James',
         'Beth', 
         'John', 
         'Sally',
         'Richard', 
         'Lauren',
         'Brandon', 
         'Sabrina']

ages = np.random.randint(18,35,len(names)) # Random Ages
my_people = {
    'names': names,
    'ages': ages
}
print(my_people)

{'names': ['Alice', 'Bob', 'James', 'Beth', 'John', 'Sally', 'Richard', 'Lauren', 'Brandon', 'Sabrina'], 'ages': array([18, 32, 29, 18, 26, 25, 33, 29, 20, 27])}


##### from_dict()

<p>Let's convert our not-so-useful-for-analysis dict into a Pandas dataframe. We can use the from_dict function to do this easily using Pandas:</p>

In [162]:
data = pd.DataFrame.from_dict(my_people)

print('Resulting type of data: {}'.format(type(data)))
data

Resulting type of data: <class 'pandas.core.frame.DataFrame'>


names  ages
0    Alice    18
1      Bob    32
2    James    29
3     Beth    18
4     John    26
5    Sally    25
6  Richard    33
7   Lauren    29
8  Brandon    20
9  Sabrina    27

##### read_csv()

In [163]:
marathon = pd.read_csv('files/boston_marathon2017_edited.csv', sep = ',')
marathon

10K                     15K                    20K  \
0      12/30/1899 12:30:28 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:15 AM   
1      12/30/1899 12:30:27 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:15 AM   
2      12/30/1899 12:30:29 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:16 AM   
3      12/30/1899 12:30:29 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:19 AM   
4      12/30/1899 12:30:28 AM  12/30/1899 12:45:44 AM  12/30/1899 1:01:15 AM   
...                       ...                     ...                    ...   
26405   12/30/1899 1:35:41 AM   12/30/1899 2:23:35 AM  12/30/1899 3:12:44 AM   
26406   12/30/1899 1:05:33 AM   12/30/1899 1:52:17 AM  12/30/1899 2:49:41 AM   
26407   12/30/1899 1:43:36 AM   12/30/1899 2:32:36 AM                    NaN   
26408   12/30/1899 1:27:19 AM   12/30/1899 2:17:17 AM  12/30/1899 3:11:40 AM   
26409   12/30/1899 1:17:12 AM   12/30/1899 2:00:10 AM  12/30/1899 2:58:55 AM   

                         25K                    30K                    35K  \
0      12/30/1899 1:16:59 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:19 AM   
1      12/30/1899 1:16:59 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:19 AM   
2      12/30/1899 1:17:00 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:31 AM   
3      12/30/1899 1:17:00 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:58 AM   
4      12/30/1899 1:16:59 AM  12/30/1899 1:33:01 AM  12/30/1899 1:48:41 AM   
...                      ...                    ...                    ...   
26405  12/30/1899 4:12:06 AM  12/30/1899 5:03:08 AM  12/30/1899 5:55:18 AM   
26406  12/30/1899 3:50:19 AM  12/30/1899 4:50:01 AM  12/30/1899 5:53:48 AM   
26407  12/30/1899 4:15:21 AM  12/30/1899 5:06:37 AM  12/30/1899 6:00:33 AM   
26408  12/30/1899 4:06:10 AM  12/30/1899 5:07:09 AM  12/30/1899 6:06:07 AM   
26409  12/30/1899 4:27:14 AM  12/30/1899 5:37:13 AM  12/30/1899 6:39:07 AM   

                         40K                      5K  Age      Bib  ...  \
0      12/30/1899 2:02:53 AM  12/30/1899 12:15:25 AM   24     11.0  ...   
1      12/30/1899 2:03:14 AM  12/30/1899 12:15:24 AM   30     17.0  ...   
2      12/30/1899 2:03:38 AM  12/30/1899 12:15:25 AM   25     23.0  ...   
3      12/30/1899 2:04:35 AM  12/30/1899 12:15:25 AM   32     21.0  ...   
4      12/30/1899 2:05:00 AM  12/30/1899 12:15:25 AM   31      9.0  ...   
...                      ...                     ...  ...      ...  ...   
26405  12/30/1899 6:46:57 AM  12/30/1899 12:46:44 AM   61  25166.0  ...   
26406  12/30/1899 6:54:21 AM  12/30/1899 12:32:03 AM   25  25178.0  ...   
26407  12/30/1899 6:54:38 AM  12/30/1899 12:53:11 AM   57  27086.0  ...   
26408  12/30/1899 6:56:08 AM  12/30/1899 12:40:34 AM   64  25268.0  ...   
26409  12/30/1899 7:41:23 AM  12/30/1899 12:39:36 AM   48  25266.0  ...   

       Division Gender                   Half  M/F                      Name  \
0             1      1  12/30/1899 1:04:35 AM    M           Kirui, Geoffrey   
1             2      2  12/30/1899 1:04:35 AM    M               Rupp, Galen   
2             3      3  12/30/1899 1:04:36 AM    M             Osako, Suguru   
3             4      4  12/30/1899 1:04:45 AM    M          Biwott, Shadrack   
4             5      5  12/30/1899 1:04:35 AM    M            Chebet, Wilson   
...         ...    ...                    ...  ...                       ...   
26405       344  11972  12/30/1899 3:23:31 AM    F  Steinbach, Paula Eyvonne   
26406      4774  14436  12/30/1899 3:00:26 AM    M        Avelino, Andrew R.   
26407       698  11973  12/30/1899 3:36:24 AM    F           Hantel, Johanna   
26408      1043  14437  12/30/1899 3:22:30 AM    M              Reilly, Bill   
26409      2553  14438  12/30/1899 3:08:16 AM    M             Rigsby, Scott   

      Number of Records          Official Time  Overall  \
0                     1  12/30/1899 2:09:37 AM        1   
1                     1  12/30/1899 2:09:58 AM        2   
2                     1  12/30/1899 2:10:28 AM        3   
3                     1  12/30

### In-Class Exercise #1 - Read in Boston Red Sox Hitting Data <br>
<p>Use the pandas read_csv() method to read in the statistics from the two files yesterday.</p>

In [164]:
bos17 = pd.read_csv('files/redsox_2017_hitting.txt', sep =',')
bos18 = pd.read_csv('files/redsox_2018_hitting.txt', sep =',')

bos18

Rk Pos                Name  Age    G   PA   AB    R    H  2B  ...    OBP  \
0    1   C          Sandy Leon   29   89  288  265   30   47  12  ...  0.232   
1    2  1B      Mitch Moreland   32  124  459  404   57   99  23  ...  0.325   
2    3  2B       Eduardo Nunez   31  127  502  480   56  127  23  ...  0.289   
3    4  SS     Xander Bogaerts   25  136  580  513   72  148  45  ...  0.360   
4    5  3B       Rafael Devers   21  121  490  450   59  108  24  ...  0.298   
5    6  LF   Andrew Benintendi   23  148  661  579  103  168  41  ...  0.366   
6    7  CF  Jackie Bradley Jr.   28  144  535  474   76  111  33  ...  0.314   
7    8  RF        Mookie Betts   25  136  614  520  129  180  47  ...  0.438   
8    9  DH       J.D. Martinez   30  150  649  569  111  188  37  ...  0.402   
9   10  MI          Brock Holt   30  109  367  321   41   89  18  ...  0.362   
10  11   C   Christian Vazquez   27   80  269  251   24   52  10  ...  0.257   
11  12  UT       Blake Swihart   26   82  207  192   28   44  10  ...  0.285   
12  13  UT      Hanley Ramirez   34   44  195  177   25   45   7  ...  0.313   
13  14  1B        Steve Pearce   35   50  165  136   19   38   8  ...  0.394   
14  15  2B         Ian Kinsler   36   37  143  132   17   32   6  ...  0.294   
15  16  SS         Tzu-Wei Lin   24   37   73   65   15   16   6  ...  0.329   
16  17  UT          Sam Travis   24   19   38   36    5    8   3  ...  0.263   
17  18  UT    Brandon Phillips   37    9   27   23    4    3   0  ...  0.259   
18  19  2B      Dustin Pedroia   34    3   13   11    1    1   0  ...  0.231   
19  20   C          Dan Butler   31    2    7    6    0    1   0  ...  0.143   

      SLG    OPS  OPS+   TB  GDP  HBP  SH  SF  IBB  
0   0.279  0.511    37   74    6    4   3   1    0  
1   0.433  0.758   102  175   12    0   0   5    2  
2   0.388  0.677    81  186   17    2   1   3    0  
3   0.522  0.883   135  268   14    6   0   6    4  
4   0.433  0.731    94  195    9    0   0   2    6  
5   0.465  0.830   123  269    9    2   2   7    1  
6   0.403  0.717    92  191    6   11   0   4    3  
7   0.640  1.078   186  333    5    8   0   5    8  
8   0.629  1.031   173  358   19    4   0   7   11  
9   0.411  0.774   109  132    7    7   0   2    2  
10  0.283  0.540    46   71    5    4   1   0    1  
11  0.328  0.613    65   63    4    0   0   0    0  
12  0.395  0.708    90   70    9    2   0   2    0  
13  0.507  0.901   141   69    1    5   0   2    0  
14  0.311  0.604    64   41    5    0   0   1    0  
15  0.415  0.744   100   27    0    0   0   0    0  
16  0.389  0.652    73   14    1    0   0   0    0  
17  0.261  0.520    42    6    1    0   0   0    0  
18  0.091  0.322    -7    1    0    0   0   0    0  
19  0.167  0.310   -17    1    0    0   0   1    0  

[20 rows x 28 columns]

### Accessing Data <br>

##### Indexing

<p>You can directly select a column of a dataframe just like you would a dict. The result is a Pandas 'Series' object.</p>

In [165]:
# Even though they are more complex series objects, they still support the behavior of the underlying NumPy arrays
data_ages = data['ages']
print(type(data_ages))

print(data_ages)

 # key first, row index second
print(data_ages[3])

print(data['names'][0])


<class 'pandas.core.series.Series'>
0    18
1    32
2    29
3    18
4    26
5    25
6    33
7    29
8    20
9    27
Name: ages, dtype: int64
18
Alice


##### df.loc

<p>Along the horizontal dimension, rows of Pandas DataFrames are Row objects. You will notice there is a third column present in the DataFrame - this is the $\textit{index}$. It is automatically generated as a row number, but can be reassigned to a column of your choice using the DataFrame.set_index(colname) method. We can use it to access particular Pandas $\textit{rows}$, which are also Series objects:</p>

In [166]:
print(data)
my_row = data.loc[0]
print("\n Location of first row of data:",my_row)
# index after assinging the variable with .loc

# indexing without defining variable
print("\n",data.loc[0]['names'])

     names  ages
0    Alice    18
1      Bob    32
2    James    29
3     Beth    18
4     John    26
5    Sally    25
6  Richard    33
7   Lauren    29
8  Brandon    20
9  Sabrina    27

 Location of first row of data: names    Alice
ages        18
Name: 0, dtype: object

 Alice


##### keys()

In [167]:
# A series can support dict-like features when it's used as a DataFrame row. 
my_row.keys()

Index(['names', 'ages'], dtype='object')

##### Slicing a DataFrame

In [168]:
 # slicing allowed
print(data[2:5])

# Slice for one item
print(data[2:3])
# The above syntax will work, but be aware, passing as single int will be interpreted as a column key
# df[2] # single indexing not allowed

   names  ages
2  James    29
3   Beth    18
4   John    26
   names  ages
2  James    29


### Built-In Methods <br>

<p>These are methods that are frequently used when using Pandas to make your life easier. It is possible to spend a whole week simply exploring the built-in functions supported by DataFrames in Pandas. Here however, we will simply highlight a few ones that might be useful, to give you an idea of what's possible out of the box with Pandas:</p>

##### head()

In [169]:
# can specify the number of rows to be displayed from the header
data.head()

names  ages
0  Alice    18
1    Bob    32
2  James    29
3   Beth    18
4   John    26

##### tail()

In [170]:
# can specify the number of rows to be displayed from the footer
data.tail()

names  ages
5    Sally    25
6  Richard    33
7   Lauren    29
8  Brandon    20
9  Sabrina    27

##### shape

In [171]:
# The dataframe has a shape property, just like a NumPy matrix. 
# print(df.shape)
print("Shape of DataFrame is: {}".format(data.shape))

# It also has an overall length property corresponding to the number of rows.
print("The DataFrame has a length of: {}".format(len(data)))

Shape of DataFrame is: (10, 2)
The DataFrame has a length of: 10


##### describe() <br>
<p>Probably one of the most important methods to understand...</p>

In [172]:
# Collect summary statistics in one line
data.describe(include ='all')

names      ages
count        10  10.00000
unique       10       NaN
top     Richard       NaN
freq          1       NaN
mean        NaN  25.70000
std         NaN   5.45792
min         NaN  18.00000
25%         NaN  21.25000
50%         NaN  26.50000
75%         NaN  29.00000
max         NaN  33.00000

##### sort_values()

In [173]:
# Sort based on many labels, with left-to-right priority
data.sort_values('ages',kind='mergesort')

print(data)

     names  ages
0    Alice    18
1      Bob    32
2    James    29
3     Beth    18
4     John    26
5    Sally    25
6  Richard    33
7   Lauren    29
8  Brandon    20
9  Sabrina    27


##### .columns

In [174]:
# will show all cols headers
data.columns

Index(['names', 'ages'], dtype='object')

### In-Class Exercise #2 - Describe & Sort Boston Red Sox Hitting Data <br>
<p>Take the data that you read in earlier from the Red Sox csv's and use the describe method to understand the data better. Compare the two years and decide which team is having the better year. Then sort the values based on Batting Average.</p>

In [175]:
print("Boston 2017 Hitting Avg: \n {}".format(bos17['BA'].describe()['mean']))
print("\n Boston 2018 Hitting Avg: \n {}".format(bos18['BA'].describe()['mean']))

Boston 2017 Hitting Avg: 
 0.27843478260869564

 Boston 2018 Hitting Avg: 
 0.23795000000000002


### Filtration <br>
<p>Let's look at how to filter dataframes for rows that fulfill a specific conditon.</p>

##### Conditionals

In [176]:
# Conditional boolean dataframe
data['ages'] >= 25

0    False
1     True
2     True
3    False
4     True
5     True
6     True
7     True
8    False
9     True
Name: ages, dtype: bool

##### Subsetting

In [177]:
# exactly like numpy
data[data['ages'] >=25]

names  ages
1      Bob    32
2    James    29
4     John    26
5    Sally    25
6  Richard    33
7   Lauren    29
9  Sabrina    27

### Column Transformations <br>
<p>Rarely, if ever, will the columns in the original raw dataframe read from CSV or database table be the ones you actually need for your analysis. You will spend lots of time constantly transforming columns or groups of columns using general computational operations to produce new ones that are functions of the old ones. Pandas has full support for this: Consider the following dataframe containing membership term and renewal number for a group of customers:</p>

In [178]:
# Generate some fake data
customer_id = np.random.randint(100,1100,10)
renewal_nbr = np.random.randint(0,10,10)
customer_dict = { 1: 0.5, 0: 1}
term_in_years = [customer_dict[key] for key in np.random.randint(0,2,10)]

# Create a dict - not usually how you would do this
random_data = {
    'customer_id':customer_id,
    'renewal_nbr': renewal_nbr,
    'term_in_years': term_in_years
}
# Build the dataframe
customers = pd.DataFrame.from_dict(random_data)
customers



customer_id  renewal_nbr  term_in_years
0          640            5            0.5
1          242            0            1.0
2          971            1            0.5
3          471            1            1.0
4          568            5            1.0
5          649            1            1.0
6          379            5            0.5
7          877            2            0.5
8          654            9            0.5
9          600            4            1.0

##### Generating a New Column w/Data

In [179]:
customers['customer_tenure'] = customers['renewal_nbr'] * customers['term_in_years']
customers.head(10)



customer_id  renewal_nbr  term_in_years  customer_tenure
0          640            5            0.5              2.5
1          242            0            1.0              0.0
2          971            1            0.5              0.5
3          471            1            1.0              1.0
4          568            5            1.0              5.0
5          649            1            1.0              1.0
6          379            5            0.5              2.5
7          877            2            0.5              1.0
8          654            9            0.5              4.5
9          600            4            1.0              4.0

##### User Defined Function

<p>If what you want to do to a column that can't be represented by simple mathematical operations, you can write your own $\textit{user defined function}$ with the full customizability available in Python and any external Python packages, then map it directly onto a column. Let's add some ages to our customer dataframe, and then classify them into our custom defined grouping scheme:</p>

In [180]:
# use .apply to map over dataframe
customers['ages'] = np.random.randint(18,70,10)

def make_age_groups(age):
    if 10 <= age and 20 > age:
        return 'Teenager'
    elif 20 <= age and 35 > age:
        return 'Young Adult'
    elif 35 <= age and 65 > age:
        return 'Adult'
    elif 65 <= age:
        return 'Senior'
    
customers['age_group'] = customers['ages'].apply(make_age_groups)
customers.head(10)


customer_id  renewal_nbr  term_in_years  customer_tenure  ages    age_group
0          640            5            0.5              2.5    27  Young Adult
1          242            0            1.0              0.0    66       Senior
2          971            1            0.5              0.5    20  Young Adult
3          471            1            1.0              1.0    60        Adult
4          568            5            1.0              5.0    48        Adult
5          649            1            1.0              1.0    58        Adult
6          379            5            0.5              2.5    52        Adult
7          877            2            0.5              1.0    67       Senior
8          654            9            0.5              4.5    40        Adult
9          600            4            1.0              4.0    50        Adult

<p>As a last example I'll show here how you would use a lambda function to create a UDF that depends on $\textit{more than one}$ column:</p>

<li>UDF = User Defined Function</li>

In [181]:
#Axis for apply can only be 1 or 0 -- 1 being the X axis 0 being the Y axis
def make_loyalty_age_groups(row):
    age = row['ages']
    tenure = row['customer_tenure']
    
    if 10 <= age and 20 > age:
        return 'Teenager'
    elif 20 <= age and 35 > age:
        return 'Young Adult'
    elif 35 <= age and 65 > age:
        return 'Adult'
    elif 65 <= age:
        age_group = 'Senior'
    
    if tenure > 2.0:
        loyalty_age_group = 'loyal ' + age_group
    else:
        loyalty_age_group = 'New ' + age_group
    
    return loyalty_age_group
    
customers['loyalty_age_group'] = customers.apply(make_loyalty_age_groups, axis = 1)
customers.head(10)


customer_id  renewal_nbr  term_in_years  customer_tenure  ages  \
0          640            5            0.5              2.5    27   
1          242            0            1.0              0.0    66   
2          971            1            0.5              0.5    20   
3          471            1            1.0              1.0    60   
4          568            5            1.0              5.0    48   
5          649            1            1.0              1.0    58   
6          379            5            0.5              2.5    52   
7          877            2            0.5              1.0    67   
8          654            9            0.5              4.5    40   
9          600            4            1.0              4.0    50   

     age_group loyalty_age_group  
0  Young Adult       Young Adult  
1       Senior        New Senior  
2  Young Adult       Young Adult  
3        Adult             Adult  
4        Adult             Adult  
5        Adult             Adult  
6        Adult             Adult  
7       Senior        New Senior  
8        Adult             Adult  
9        Adult             Adult

### In-Class Exercise #3 - Create Your Own UDF <br>
<p>Using the Boston Red Sox data, create your own UDF which creates a new column called 'All-Star' and puts every player with either a batting average over .280 or an on base percentage of over .360 with a result of 'Yes' in the column and 'No' if not.</p>

In [182]:
def check_all_star(row):
    avg = row['BA']
    obp = row['OBP']
    
    if avg > 0.28 or obp > 0.35:
        return 'Yes'
    else:
        return 'No'
    
bos18['AllStar'] = bos18.apply(check_all_star,axis=1)
bos18[['Name','BA','OBP','AllStar']]

# bos18[bos18['AllStar'] == 'Yes']

Name     BA    OBP AllStar
0           Sandy Leon  0.177  0.232      No
1       Mitch Moreland  0.245  0.325      No
2        Eduardo Nunez  0.265  0.289      No
3      Xander Bogaerts  0.288  0.360     Yes
4        Rafael Devers  0.240  0.298      No
5    Andrew Benintendi  0.290  0.366     Yes
6   Jackie Bradley Jr.  0.234  0.314      No
7         Mookie Betts  0.346  0.438     Yes
8        J.D. Martinez  0.330  0.402     Yes
9           Brock Holt  0.277  0.362     Yes
10   Christian Vazquez  0.207  0.257      No
11       Blake Swihart  0.229  0.285      No
12      Hanley Ramirez  0.254  0.313      No
13        Steve Pearce  0.279  0.394     Yes
14         Ian Kinsler  0.242  0.294      No
15         Tzu-Wei Lin  0.246  0.329      No
16          Sam Travis  0.222  0.263      No
17    Brandon Phillips  0.130  0.259      No
18      Dustin Pedroia  0.091  0.231      No
19          Dan Butler  0.167  0.143      No

### Aggregations <br>
<p>The raw data plus some transformations is generally only half the story. Your objective is to extract actual insights and actionable conclusions from the data, and that means reducing it from potentially billions of rows to some summary statistics via aggregation functions.</p>

##### groupby() <br>
<p>The .groupby() function is in some ways a 'master' aggregation.</p> 

<p>Data tables will usually reserve one column as a primary key - that is, a column for which each row has a unique value. This is to facilitate access to the exact rows of a data table that a user wants to view. The other columns will often have repeated values, such as the age groups in the above examples. We can use these columns to explore the data using the Pandas API:</p>

In [183]:
# also introducing .count() here, exact same as to how it's used in SQL
customers.groupby('age_group', as_index = False).count().head()

age_group  customer_id  renewal_nbr  term_in_years  customer_tenure  \
0        Adult            6            6              6                6   
1       Senior            2            2              2                2   
2  Young Adult            2            2              2                2   

   ages  loyalty_age_group  
0     6                  6  
1     2                  2  
2     2                  2

##### Type of groupby()

<p>The result is a new dataframe, the columns of which all contain the counts of the grouped field. Notice the type of a grouped dataframe:</p>

In [184]:
type(customers.groupby('age_group', as_index = False))

pandas.core.groupby.generic.DataFrameGroupBy

<p>This is because simply grouping data doesn't quite make sense without an aggregation function like count() to pair with. In this case, we're counting occurances of the grouped field, but that's not all we can do. We can take averages, standard deviations, mins, maxes and much more! Let's see how this works a bit more:</p>

##### mean()

In [185]:
# mean = average
customers.groupby('age_group', as_index = False).mean()[['age_group','ages']]

age_group       ages
0        Adult  51.333333
1       Senior  66.500000
2  Young Adult  23.500000

##### groupby() w/Multiple Columns

<p>We end up with the average age of the groups in the last column, the average tenure in the tenure column, and so on and so forth. You can even split the groups more finely by passing a list of columns to group by:</p>

In [186]:
customers.groupby('age_group', as_index = False).mean()[['age_group','ages']]



age_group       ages
0        Adult  51.333333
1       Senior  66.500000
2  Young Adult  23.500000

##### drop_duplicates()

<p>Drops all duplicates from the current dataframe</p>

In [187]:
#cust_copy = customers

result = cust_copy.drop_duplicates(['age_group'])
result

customer_id  renewal_nbr  term_in_years  customer_tenure  ages  \
0          769            1            1.0              1.0    25   
2          762            1            0.5              0.5    68   
5          319            9            1.0              9.0    55   

     age_group loyalty_age_group  
0  Young Adult       Young Adult  
2       Senior        New Senior  
5        Adult             Adult

In [188]:
customers.describe(include = 'all')

customer_id  renewal_nbr  term_in_years  customer_tenure       ages  \
count     10.000000    10.000000      10.000000        10.000000  10.000000   
unique          NaN          NaN            NaN              NaN        NaN   
top             NaN          NaN            NaN              NaN        NaN   
freq            NaN          NaN            NaN              NaN        NaN   
mean     605.100000     3.300000       0.750000         2.200000  48.800000   
std      214.915203     2.790858       0.263523         1.782632  15.746605   
min      242.000000     0.000000       0.500000         0.000000  20.000000   
25%      495.250000     1.000000       0.500000         1.000000  42.000000   
50%      620.000000     3.000000       0.750000         1.750000  51.000000   
75%      652.750000     5.000000       1.000000         3.625000  59.500000   
max      971.000000     9.000000       1.000000         5.000000  67.000000   

       age_group loyalty_age_group  
count         10                10  
unique         3                 3  
top        Adult             Adult  
freq           6                 6  
mean         NaN               NaN  
std          NaN               NaN  
min          NaN               NaN  
25%          NaN               NaN  
50%          NaN               NaN  
75%          NaN               NaN  
max          NaN               NaN

<p>Thus the groupby operation allows you to rapidly make summary observations about the state of your entire dataset at flexible granularity. In one line above, we actually did something very complicated - that's the power of the dataframe. In fact, the process often consists of several iterative groupby operations, each revealing greater insight than the last - if you don't know where to start with a dataset, try a bunch of groupbys!</p>

### Homework Excersise #4 - Find the Mean for the Red Sox <br>
<p>For both 2017 and 2018, find the mean of the batting averages</p>

In [189]:
bos17['Team'] = 'BOS'

bos18['Team'] = 'BOS'

print(bos17.groupny('Team', as_index = False).sum()[['Team','HR','RBI']])
print(bos18.groupny('Team', as_index = False).sum()[['Team','HR','RBI']])


AttributeError: 'DataFrame' object has no attribute 'groupny'